Userland/Libraries/LibVideo/VP9/Decoder.cpp

```diff
@@ -907,71 +907,178 @@ DecoderErrorOr<void> Decoder::predict_inter_block(u8 plane, BlockContext const&
+            return {};
         }
-    }
 
-    for (auto row = 0u; row < height; row++) {
-        for (auto column = 0u; column < width; column++) {
-            auto samples_start = (offset_scaled_block_y & 15) + static_cast<i32>(scaled_step_y * row);
-            auto const* scan_column = &intermediate_buffer_at(samples_start >> 4, column);
-            auto const* subpel_filters_for_samples = subpel_filters[block_context.interpolation_filter][samples_start & 15];
-
-            i32 accumulated_samples = 0;
-            for (auto t = 0u; t < 8u; t++) {
-                auto sample = *scan_column;
-                accumulated_samples += subpel_filters_for_samples[t] * sample;
-                scan_column += width;
+        auto vertical_convolution_unscaled = [](auto bit_depth, auto* destination, auto width, auto height, auto const* source, auto source_stride, auto filter, auto subpixel_y) {
+            auto const source_end_skip = source_stride - width;
+
+            for (auto row = 0u; row < height; row++) {
+                for (auto column = 0u; column < width; column++) {
+                    auto const* scan_column = source;
+                    i16 accumulated_samples = 0;
+                    for (auto t = 0; t < 8; t++) {
+                        auto sample = *scan_column;
+                        accumulated_samples += subpel_filters[filter][subpixel_y][t] * sample;
+                        scan_column += source_stride;
+                    }
+                    *destination = clip_1(bit_depth, rounded_right_shift(accumulated_samples, 7));
+                    source++;
+                    destination++;
+                }
+                source += source_end_skip;
             }
-            block_buffer_at(row, column) = clip_1(block_context.frame_context.color_config.bit_depth, rounded_right_shift(accumulated_samples, 7));
+        };
+
+        if (copy_x) {
+            vertical_convolution_unscaled(bit_depth, block_buffer.data(), width, height, reference_start - (sample_offset * reference_frame_width), reference_frame_width, block_context.interpolation_filter, reference_subpixel_y);
+            return {};
         }
+
+        horizontal_convolution_unscaled(bit_depth, intermediate_buffer.data(), width, intermediate_height, reference_start - (sample_offset * reference_frame_width), reference_frame_width, block_context.interpolation_filter, reference_subpixel_x);
+        vertical_convolution_unscaled(bit_depth, block_buffer.data(), width, height, intermediate_buffer.data(), width, block_context.interpolation_filter, reference_subpixel_y);
+        return {};
     }
 
+    // NOTE: Accumulators below are 32-bit to allow high bit-depth videos to decode without overflows.
+    //       These should be changed when the accumulators above are.
```

Comment:
maybe using a type alias would be more explicit, thus clearer and more maintainable?

Reply:
That's something that I'll have to work into this at a later date, since I have to experiment to see what kind of benefit using smaller integer sizes has on performance when all the hot code is written in an efficient manner.

Reply:
it seems you could start from one type alias — the one mentioned in the comments. yet i understand this is somewhat out of scope and its better to have such comment than not having it. all understood, consider this comment resolved! 🙇🏻‍♂️